### ISSUE TITLE
task: Train and test t2s under decoder only framework for ichigo v0.5

### ISSUE THREAD

**Issue by tikikun** [janhq/ichigo#139]
Since ichigo v0.5 will support additional language that will make the traditional t2s obsolete. This is a good chance to introduce a t2s framework that we have full control over.

# What needed to be done:
-  [ ] Train t2s on a decoder model to test feasibility (we already have the data) under ichigo v0.4
-  [ ] Train t2s on decoder model on new quantizer after #120 is done

**Comment by bachvudinh** [comment]
## Goal
- The current synthetic data pipeline for Ichigo Training is constrained by text-to-speech model performance. As we expand to multilingual support, the text-to-speech component becomes a significant bottleneck. To address this limitation, we need to develop and train our own multilingual text-to-semantic model for synthetic data generation.

## Methodology
- After extensive hyperparameter tuning, the WhisperSpeech text-to-semantic model proved inadequate for our synthetic data pipeline. The model's output resulted in incompatible sound tokens that broke Ichigo's comprehension capabilities.
To address this, we propose developing a custom text-to-semantic model based on a decoder-only architecture (similar to Qwen 2.5) with <3B parameters. This model will leverage knowledge transfer from WhisperVQ while maintaining a more efficient architecture that better aligns with our use case.
- To develop our custom text-to-semantic model, I processed 10k English samples from the MLS Eng 10k (2.42M samples) dataset by tokenizing the raw speech using WhisperVQ to extract semantic tokens. I also add a special task token **<|text_to_semantic|>** in user turn. Here is a sample from [Instruction data](https://huggingface.co/datasets/homebrewltd/instruction-convert-audio-whispervq):
```
[ { "content": "<|text_to_sementic|>he telegraphed to general pemberton that he had learned sherman was between them with four divisions at clinton saying that it was important to reestablish communications that pemberton might be reenforced and directing him to come up in sherman's rear at once", "role": "user" }, { "content": "<|sound_start|><|sound_0209|><|sound_0134|><|sound_0134|><|sound_0134|><|sound_0241|><|sound_0222|><|sound_0239|><|sound_0329|><|sound_0197|><|sound_0115|><|sound_0409|><|sound_0196|><|sound_0196|><|sound_0196|><|sound_0196|><|sound_0196|><|sound_0235|><|sound_0487|><|sound_0487|><|sound_0459|><|sound_0459|><|sound_0405|><|sound_end|>", "role": "assistant" } ]
```
- I then modified Qwen 2.5 0.5B by introducing a new task token **<|text_to_semantic|>** and incorporating 515 sound tokens into its vocabulary. The training data was structured as instruction-based samples designed to teach the model text-to-semantic token conversion.  Following Qwen's methodology, we integrated control tokens into the embedding layer without tokenizer modification which, according to Qwen Authors, aim at optimizing training performance, we adjusted the embedding layer and LM head dimensions by a factor of 128, resulting in a final embedding size of 152,192 tokens. 

## Hyperparams
- Enable both activation checkpointing and activation offloading to reduce vram usage.
- It took ~ 100 hours on 8xA6000

Parameter | Value
--- | ---
Lora Rank | 256
Lora Alpha | 512
Epoch | 1
Global batch size | 384
Learning Rate | 1.5e-4
Learning Scheduler | Cosine
Optimizer | AdamW Fused
Warmup Steps | 63
Weight Decay | 0.01
Gradient Checkpointing | Full
Max length | 512
Precision | bf16
Clip_grad_norm | 1.0

## Results
- Loss Curve: 
![image](https://github.com/user-attachments/assets/f91a7957-54a9-4de0-a555-565123b57cd8)

- MMLU score:

- MMLU pro score:

- some extra benchmark(adding soon):

## Learnings


## Quicklinks
- Data: https://huggingface.co/datasets/homebrewltd/mixed-instruction-speech-whispervq-v4
- Checkpoints:

**Comment by bachvudinh** [comment]
![image](https://github.com/user-attachments/assets/23400663-ca7e-4fb0-b8ba-50776ab8329c)

**Comment by dan-homebrew** [comment]
Need to add more details to this issue: 

- Goal: Be able to handle any arbitrary language
    - Current Ichigo approach (WhisperVQ) is only trained for 7 languages
    - We cannot find a ASR or STT module for every new language
    - Our approach: Text to Semantic (that people can listen to)
